File: cpp-package/example/resnet.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 */
#include <map>
#include <string>
#include <fstream>
#include <vector>
#include <cstdlib>
#include "utils.h"
#include "mxnet-cpp/MxNetCpp.h"

using namespace mxnet::cpp;

Symbol ConvolutionNoBias(const std::string& symbol_name,
                         Symbol data,
                         Symbol weight,
                         Shape kernel,
                         int num_filter,
                         Shape stride = Shape(1, 1),
                         Shape dilate = Shape(1, 1),
                         Shape pad = Shape(0, 0),
                         int num_group = 1,
                         int64_t workspace = 512) {
  return Operator("Convolution")
      .SetParam("kernel", kernel)
      .SetParam("num_filter", num_filter)
      .SetParam("stride", stride)
      .SetParam("dilate", dilate)
      .SetParam("pad", pad)
      .SetParam("num_group", num_group)
      .SetParam("workspace", workspace)
      .SetParam("no_bias", true)
      .SetInput("data", data)
      .SetInput("weight", weight)
      .CreateSymbol(symbol_name);
}

Symbol getConv(const std::string & name, Symbol data,
               int  num_filter,
               Shape kernel, Shape stride, Shape pad,
               bool with_relu,
               mx_float bn_momentum) {
  Symbol conv_w(name + "_w");
  Symbol conv = ConvolutionNoBias(name, data, conv_w,
                                  kernel, num_filter, stride, Shape(1, 1),
                                  pad, 1, 512);

  Symbol gamma(name + "_gamma");
  Symbol beta(name + "_beta");
  Symbol mmean(name + "_mmean");
  Symbol mvar(name + "_mvar");

  Symbol bn = BatchNorm(name + "_bn", conv, gamma,
                        beta, mmean, mvar, 2e-5, bn_momentum, false);

  if (with_relu) {
    return Activation(name + "_relu", bn, "relu");
  } else {
    return bn;
  }
}

Symbol makeBlock(const std::string & name, Symbol data, int num_filter,
                 bool dim_match, mx_float bn_momentum) {
  Shape stride;
  if (dim_match) {
    stride = Shape(1, 1);
  } else {
    stride = Shape(2, 2);
  }

  Symbol conv1 = getConv(name + "_conv1", data, num_filter,
                         Shape(3, 3), stride, Shape(1, 1),
                         true, bn_momentum);

  Symbol conv2 = getConv(name + "_conv2", conv1, num_filter,
                         Shape(3, 3), Shape(1, 1), Shape(1, 1),
                         false, bn_momentum);

  Symbol shortcut;

  if (dim_match) {
    shortcut = data;
  } else {
    Symbol shortcut_w(name + "_proj_w");
    shortcut = ConvolutionNoBias(name + "_proj", data, shortcut_w,
                                 Shape(2, 2), num_filter,
                                 Shape(2, 2), Shape(1, 1), Shape(0, 0),
                                 1, 512);
  }

  Symbol fused = shortcut + conv2;
  return Activation(name + "_relu", fused, "relu");
}

Symbol getBody(Symbol data, int num_level, int num_block, int num_filter, mx_float bn_momentum) {
  for (int level = 0; level < num_level; level++) {
    for (int block = 0; block < num_block; block++) {
      data = makeBlock("level" + std::to_string(level + 1) + "_block" + std::to_string(block + 1),
                       data, num_filter * (std::pow(2, level)),
                       (level == 0 || block > 0), bn_momentum);
    }
  }
  return data;
}

Symbol ResNetSymbol(int num_class, int num_level = 3, int num_block = 9,
                    int num_filter = 16, mx_float bn_momentum = 0.9,
                    mxnet::cpp::Shape pool_kernel = mxnet::cpp::Shape(8, 8)) {
  // data and label
  Symbol data = Symbol::Variable("data");
  Symbol data_label = Symbol::Variable("data_label");

  Symbol gamma("gamma");
  Symbol beta("beta");
  Symbol mmean("mmean");
  Symbol mvar("mvar");

  Symbol zscore = BatchNorm("zscore", data, gamma,
                            beta, mmean, mvar, 0.001, bn_momentum);

  Symbol conv = getConv("conv0", zscore, num_filter,
                        Shape(3, 3), Shape(1, 1), Shape(1, 1),
                        true, bn_momentum);

  Symbol body = getBody(conv, num_level, num_block, num_filter, bn_momentum);

  Symbol pool = Pooling("pool", body, pool_kernel, PoolingPoolType::kAvg);

  Symbol flat = Flatten("flatten", pool);

  Symbol fc_w("fc_w"), fc_b("fc_b");
  Symbol fc = FullyConnected("fc", flat, fc_w, fc_b, num_class);

  return SoftmaxOutput("softmax", fc, data_label);
}

NDArray ResizeInput(NDArray data, const Shape new_shape) {
  NDArray pic = data.Reshape(Shape(0, 1, 28, 28));
  NDArray pic_1channel;
  Operator("_contrib_BilinearResize2D")
    .SetParam("height", new_shape[2])
    .SetParam("width", new_shape[3])
    (pic).Invoke(pic_1channel);
  NDArray output;
  Operator("tile")
    .SetParam("reps", Shape(1, 3, 1, 1))
    (pic_1channel).Invoke(output);
  return output;
}

int main(int argc, char const *argv[]) {
  int max_epoch = argc > 1 ? strtol(argv[1], NULL, 10) : 100;
  float learning_rate = 1e-4;
  float weight_decay = 1e-4;

  TRY
  auto resnet = ResNetSymbol(10);
  std::map<std::string, NDArray> args_map;
  std::map<std::string, NDArray> aux_map;

  /*context*/
  auto ctx = Context::cpu();
  int num_gpu;
  MXGetGPUCount(&num_gpu);
  int batch_size = 8;
#if !MXNET_USE_CPU
  if (num_gpu > 0) {
    ctx = Context::gpu();
    batch_size = 32;
  }
#endif

  const Shape data_shape = Shape(batch_size, 3, 224, 224),
              label_shape = Shape(batch_size);
  args_map["data"] = NDArray(data_shape, ctx);
  args_map["data_label"] = NDArray(label_shape, ctx);
  resnet.InferArgsMap(ctx, &args_map, args_map);

  std::vector<std::string> data_files = { "./data/mnist_data/train-images-idx3-ubyte",
                                          "./data/mnist_data/train-labels-idx1-ubyte",
                                          "./data/mnist_data/t10k-images-idx3-ubyte",
                                          "./data/mnist_data/t10k-labels-idx1-ubyte"
                                        };

  auto train_iter =  MXDataIter("MNISTIter");
  if (!setDataIter(&train_iter, "Train", data_files, batch_size)) {
    return 1;
  }

  auto val_iter = MXDataIter("MNISTIter");
  if (!setDataIter(&val_iter, "Label", data_files, batch_size)) {
    return 1;
  }

  // initialize parameters
  Xavier xavier = Xavier(Xavier::gaussian, Xavier::in, 2);
  for (auto &arg : args_map) {
    xavier(arg.first, &arg.second);
  }

  Optimizer* opt = OptimizerRegistry::Find("sgd");
  opt->SetParam("lr", learning_rate)
     ->SetParam("wd", weight_decay)
     ->SetParam("momentum", 0.9)
     ->SetParam("rescale_grad", 1.0 / batch_size)
     ->SetParam("clip_gradient", 10);

  auto *exec = resnet.SimpleBind(ctx, args_map);
  auto arg_names = resnet.ListArguments();

  // Create metrics
  Accuracy train_acc, val_acc;
  LogLoss logloss_train, logloss_val;
  for (int epoch = 0; epoch < max_epoch; ++epoch) {
    LG << "Epoch: " << epoch;
    train_iter.Reset();
    train_acc.Reset();
    int iter = 0;
    while (train_iter.Next()) {
      auto data_batch = train_iter.GetDataBatch();
      ResizeInput(data_batch.data, data_shape).CopyTo(&args_map["data"]);
      data_batch.label.CopyTo(&args_map["data_label"]);
      NDArray::WaitAll();

      exec->Forward(true);
      exec->Backward();

      for (size_t i = 0; i < arg_names.size(); ++i) {
        if (arg_names[i] == "data" || arg_names[i] == "data_label") continue;
        opt->Update(i, exec->arg_arrays[i], exec->grad_arrays[i]);
      }
      NDArray::WaitAll();
      train_acc.Update(data_batch.label, exec->outputs[0]);
      logloss_train.Reset();
      logloss_train.Update(data_batch.label, exec->outputs[0]);
      ++iter;
      LG << "EPOCH: " << epoch << " ITER: " << iter
         << " Train Accuracy: " << train_acc.Get()
         << " Train Loss: " << logloss_train.Get();
    }
    LG << "EPOCH: " << epoch << " Train Accuracy: " << train_acc.Get();

    val_iter.Reset();
    val_acc.Reset();
    iter = 0;
    while (val_iter.Next()) {
      auto data_batch = val_iter.GetDataBatch();
      ResizeInput(data_batch.data, data_shape).CopyTo(&args_map["data"]);
      data_batch.label.CopyTo(&args_map["data_label"]);
      NDArray::WaitAll();
      exec->Forward(false);
      NDArray::WaitAll();
      val_acc.Update(data_batch.label, exec->outputs[0]);
      LG << "EPOCH: " << epoch << " ITER: " << iter << " Val Accuracy: " << val_acc.Get();
      ++iter;
    }
    LG << "Validation Accuracy: " << val_acc.Get();
  }
  delete exec;
  delete opt;
  MXNotifyShutdown();
  CATCH
  return 0;
}

File: scala-package/examples/src/main/scala/org/apache/mxnetexamples/imclassification/models/Resnet.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mxnetexamples.imclassification.models

import org.apache.mxnet._

object Resnet {
  /**
    * Helper to produce individual residual unit
    */
  def residualUnit(data: Symbol, numFilter: Int, stride: Shape, dimMatch: Boolean,
                   name: String = "", bottleNeck: Boolean = true, bnMom: Float = 0.9f,
                   workspace: Int = 256, memonger: Boolean = false): Symbol = {
    val (act1, operated) = if (bottleNeck) {
      val bn1 = Symbol.api.BatchNorm(data = Some(data), fix_gamma = Some(false), eps = Some(2e-5),
        momentum = Some(bnMom), name = name + "_bn1")
      val act1: Symbol = Symbol.api.Activation(data = Some(bn1), act_type = "relu",
        name = name + "_relu1")
      val conv1 = Symbol.api.Convolution(data = Some(act1), num_filter = (numFilter * 0.25).toInt,
        kernel = Shape(1, 1), stride = Some(Shape(1, 1)), pad = Some(Shape(0, 0)),
        no_bias = Some(true), workspace = Some(workspace), name = name + "_conv1")
      val bn2 = Symbol.api.BatchNorm(data = Some(conv1), fix_gamma = Some(false),
        eps = Some(2e-5), momentum = Some(bnMom), name = name + "_bn2")
      val act2 = Symbol.api.Activation(data = Some(bn2), act_type = "relu", name = name + "_relu2")
      val conv2 = Symbol.api.Convolution(data = Some(act2), num_filter = (numFilter * 0.25).toInt,
        kernel = Shape(3, 3), stride = Some(stride), pad = Some(Shape(1, 1)),
        no_bias = Some(true), workspace = Some(workspace), name = name + "_conv2")
      val bn3 = Symbol.api.BatchNorm(data = Some(conv2), fix_gamma = Some(false),
        eps = Some(2e-5), momentum = Some(bnMom), name = name + "_bn3")
      val act3 = Symbol.api.Activation(data = Some(bn3), act_type = "relu", name = name + "_relu3")
      val conv3 = Symbol.api.Convolution(data = Some(act3), num_filter = numFilter,
        kernel = Shape(1, 1), stride = Some(Shape(1, 1)), pad = Some(Shape(0, 0)),
        no_bias = Some(true), workspace = Some(workspace), name = name + "_conv3")
      (act1, conv3)
    } else {
      val bn1 = Symbol.api.BatchNorm(data = Some(data), fix_gamma = Some(false),
        eps = Some(2e-5), momentum = Some(bnMom), name = name + "_bn1")
      val act1 = Symbol.api.Activation(data = Some(bn1), act_type = "relu", name = name + "_relu1")
      val conv1 = Symbol.api.Convolution(data = Some(act1), num_filter = numFilter,
        kernel = Shape(3, 3), stride = Some(stride), pad = Some(Shape(1, 1)),
        no_bias = Some(true), workspace = Some(workspace), name = name + "_conv1")
      val bn2 = Symbol.api.BatchNorm(data = Some(conv1), fix_gamma = Some(false),
        eps = Some(2e-5), momentum = Some(bnMom), name = name + "_bn2")
      val act2 = Symbol.api.Activation(data = Some(bn2), act_type = "relu", name = name + "_relu2")
      val conv2 = Symbol.api.Convolution(data = Some(act2), num_filter = numFilter,
        kernel = Shape(3, 3), stride = Some(Shape(1, 1)), pad = Some(Shape(1, 1)),
        no_bias = Some(true), workspace = Some(workspace), name = name + "_conv2")
      (act1, conv2)
    }
    val shortcut = if (dimMatch) {
      data
    } else {
      Symbol.api.Convolution(Some(act1), num_filter = numFilter, kernel = Shape(1, 1),
        stride = Some(stride), no_bias = Some(true), workspace = Some(workspace),
        name = name + "_sc")
    }
    operated + shortcut
  }

  /**
    * Helper for building the resnet Symbol
    */
  def resnet(units: List[Int], numStages: Int, filterList: List[Int], numClasses: Int,
             imageShape: List[Int], bottleNeck: Boolean = true, bnMom: Float = 0.9f,
             workspace: Int = 256, dtype: String = "float32", memonger: Boolean = false): Symbol = {
    assert(units.size == numStages)
    var data = Symbol.Variable("data", shape = Shape(List(4) ::: imageShape), dType = DType.Float32)
    if (dtype == "float32") {
      data = Symbol.api.identity(Some(data), "id")
    } else if (dtype == "float16") {
      data = Symbol.api.cast(Some(data), "float16")
    }
    data = Symbol.api.BatchNorm(Some(data), fix_gamma = Some(true), eps = Some(2e-5),
      momentum = Some(bnMom), name = "bn_data")
    val List(channels, height, width) = imageShape
    var body = if (height <= 32) {
      Symbol.api.Convolution(Some(data), num_filter = filterList.head, kernel = Shape(7, 7),
        stride = Some(Shape(1, 1)), pad = Some(Shape(1, 1)), no_bias = Some(true), name = "conv0",
        workspace = Some(workspace))
    } else {
      var body0 = Symbol.api.Convolution(Some(data), num_filter = filterList.head,
        kernel = Shape(3, 3), stride = Some(Shape(2, 2)), pad = Some(Shape(3, 3)),
        no_bias = Some(true), name = "conv0", workspace = Some(workspace))
      body0 = Symbol.api.BatchNorm(Some(body0), fix_gamma = Some(false), eps = Some(2e-5),
        momentum = Some(bnMom), name = "bn0")
      body0 = Symbol.api.Activation(Some(body0), act_type = "relu", name = "relu0")
      Symbol.api.Pooling(Some(body0), kernel = Some(Shape(3, 3)), stride = Some(Shape(2, 2)),
        pad = Some(Shape(1, 1)), pool_type = Some("max"))
    }
    for (((filter, i), unit) <- filterList.tail.zipWithIndex.zip(units)) {
      val stride = Shape(if (i == 0) 1 else 2, if (i == 0) 1 else 2)
      body = residualUnit(body, filter, stride, false, name = s"stage${i + 1}_unit${1}",
        bottleNeck = bottleNeck, workspace = workspace, memonger = memonger)
      for (j <- 0 until unit - 1) {
        body = residualUnit(body, filter, Shape(1, 1), true, s"stage${i + 1}_unit${j + 2}",
          bottleNeck, workspace = workspace, memonger = memonger)
      }
    }
    val bn1 = Symbol.api.BatchNorm(Some(body), fix_gamma = Some(false), eps = Some(2e-5),
      momentum = Some(bnMom), name = "bn1")
    val relu1 = Symbol.api.Activation(Some(bn1), act_type = "relu", name = "relu1")
    val pool1 = Symbol.api.Pooling(Some(relu1), global_pool = Some(true),
      kernel = Some(Shape(7, 7)), pool_type = Some("avg"), name = "pool1")
    val flat = Symbol.api.Flatten(Some(pool1))
    var fc1 = Symbol.api.FullyConnected(Some(flat), num_hidden = numClasses, name = "fc1")
    if (dtype == "float16") {
      fc1 = Symbol.api.cast(Some(fc1), "float32")
    }
    Symbol.api.SoftmaxOutput(Some(fc1), name = "softmax")
  }

  /**
    * Gets the resnet model symbol
    * @param numClasses Number of classes to classify into
    * @param numLayers Number of residual layers
    * @param imageShape The image shape as List(channels, height, width)
    * @param convWorkspace Maximum temporary workspace allowed (MB) in convolutions
    * @param dtype Type of data (float16, float32, etc) to use during computation
    * @return Model symbol
    */
  def getSymbol(numClasses: Int, numLayers: Int, imageShape: List[Int], convWorkspace: Int = 256,
                dtype: String = "float32"): Symbol = {
    val List(channels, height, width) = imageShape
    val (numStages, units, filterList, bottleNeck): (Int, List[Int], List[Int], Boolean) =
      if (height <= 28) {
        val (perUnit, filterList, bottleNeck) = if ((numLayers - 2) % 9 == 0 && numLayers > 165) {
          (List(Math.floor((numLayers - 2) / 9).toInt),
            List(16, 64, 128, 256),
            true)
        } else if ((numLayers - 2) % 6 == 0 && numLayers < 164) {
          (List(Math.floor((numLayers - 2) / 6).toInt),
            List(16, 16, 32, 64),
            false)
        } else {
          throw new Exception(s"Invalid number of layers: ${numLayers}")
        }
        val numStages = 3
        val units = (1 to numStages).map(_ => perUnit.head).toList
        (numStages, units, filterList, bottleNeck)
      } else {
        val (filterList, bottleNeck) = if (numLayers >= 50) {
          (List(64, 256, 512, 1024, 2048), true)
        } else {
          (List(64, 64, 128, 256, 512), false)
        }
        val units: List[Int] = Map(
          18 -> List(2, 2, 2, 2),
          34 -> List(3, 4, 6, 3),
          50 -> List(3, 4, 6, 3),
          101 -> List(3, 4, 23, 3),
          152 -> List(3, 8, 36, 3),
          200 -> List(3, 24, 36, 3),
          269 -> List(3, 30, 48, 8)
        ).get(numLayers) match {
          case Some(x) => x
          case None => throw new Exception(s"Invalid number of layers: ${numLayers}")
        }
        (4, units, filterList, bottleNeck)
      }
    resnet(units, numStages, filterList, numClasses, imageShape, bottleNeck,
      workspace = convWorkspace, dtype = dtype)
  }
}
